libyul/YulName.h

```diff
@@ -18,9 +18,114 @@
 
 #pragma once
 
+#include <libyul/SideEffects.h>
 #include <libyul/YulString.h>
 
+#include <fmt/format.h>
+
+#include <memory>
+#include <vector>
+#include <string>
+#include <optional>
+#include <functional>
+
 namespace solidity::yul
 {
+
+enum class LiteralKind;
+struct Block;
+
+/**
+ * Yul Name repository.
+ *
+ * A database of numerical identifiers for Yul nodes in an AST (`YulName`s). Identifiers can be
+ *
+ *  - 'defined', i.e., they are equipped with a string label, which can be retrieved by a call to `labelOf(yulName)`;
+ *  - 'derived', i.e., they don't possess a string label but have a parent yul name which is either also derived or
+ *    defined. All dependency chains of derived labels terminate in a defined label.
+ *
+ * Such derived identifiers can be introduced, e.g., during certain optimization steps. If the AST (or segments thereof)
+ * should be printed and/or the string label of identifiers should be retrieved which are still in derived state,
+ * a call to `generateLabels(identifiers)` changes the status of all derived labels to defined and generates
+ * unique labels for all identifiers provided to the method. The generated labels are based on their parents.
+ */
+class YulNameRepository
+{
+public:
+	using YulName = std::uint64_t;
```

Comment:
Since we decided that we don't want to safeguard this from being used with the wrong repository, we should at least add a warning here saying that such usage is unsafe and it's up to the user of the class to take precautions against it.